File: grind-75/cpp/1. Two Sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> valueToLocationMap;

    for (int i = 0; i < nums.size(); ++i) {
        int num = nums[i];  // current number
        int numTarget = target - num;

        // Check if complement is in map
        auto it = valueToLocationMap.find(numTarget);
        if ( it != valueToLocationMap.end()) {
            return { it->second, i};
        }

        valueToLocationMap[num] = i;
    }

    return {};
}

File: grind-75/typescript/1. Two Sum.ts
function twoSum(nums: number[], target: number): number[] {
    
    const numPosMap = new Map<number, number>();

    for (let i=0; i<nums.length; i++) {
        const num = nums[i]; // 2 | 7
        const numTargetDelta = target-num; // 9-2 -> 7 | 9-7 -> 2
        const mapEntry = numPosMap.get(numTargetDelta); // undefined | 0

        if (mapEntry !== undefined) return [ mapEntry, i ]; // ! | [ 0, 1 ]
        numPosMap.set(num, i); // add [2, 0]
    }

    return [];
};